compiler/src/iree/compiler/Dialect/Flow/Transforms/RaiseSpecialOps.cpp

```diff
@@ -27,6 +28,170 @@ namespace Flow {
+    //    1. A constant value.
+    //    2. A linalg.index.
+
+    // 1. Indexing value is a constant.
+    APInt constantIndex;
+    if (matchPattern(indexValue, m_ConstantInt(&constantIndex))) {
+      exprs.push_back(getAffineConstantExpr(constantIndex.getLimitedValue(),
+                                            rewriter.getContext()));
+      continue;
+    }
+    // 2. The indexing value is a linalg.index.
+    assert(isElementwise(linalgOp) && "Expected elementwise linalg op");
+    if (auto indexOp = indexValue.getDefiningOp<linalg::IndexOp>()) {
+      exprs.push_back(
+          getAffineDimExpr(indexOp.getDim(), rewriter.getContext()));
+      continue;
+    }
+    return failure();
+  }
+  AffineMap indexingMap = AffineMap::get(
+      /*dimCount=*/linalgOp.getNumLoops(),
+      /*symbolCount=*/0, exprs, rewriter.getContext());
+
+  // Replace the linalgOp with a new linalgOp where the source tensor is
+  // an input with the indexing map.
+  SmallVector<Value> newInputs = linalgOp.getInputs();
+  newInputs.insert(newInputs.begin(), extractOp.getTensor());
+  SmallVector<Attribute> newIndexingMaps;
+  newIndexingMaps.push_back(AffineMapAttr::get(indexingMap));
+  for (AffineMap map : linalgOp.getIndexingMapsArray())
+    newIndexingMaps.push_back(AffineMapAttr::get(map));
+
+  auto bodyBuilder = [&](OpBuilder &builder, Location loc, ValueRange args) {
+    // Create an IR mapping from old block arguements to new ones.
+    IRMapping mapper;
+    ArrayRef<BlockArgument> oldArgs = linalgOp.getBody()->getArguments();
+    // Map i^th old argument to (i + 1)^th new argument.
+    for (unsigned i = 0; i < oldArgs.size(); ++i)
+      mapper.map(oldArgs[i], args[i + 1]);
+    // Clone the body of the linalgOp.
+    for (Operation &op : linalgOp.getBody()->getOperations()) {
+      // Replace the extractOp with the first block argument.
+      if (&op == extractOp)
+        mapper.map(op.getResult(0), args[0]);
+      else
+        builder.clone(op, mapper);
+    }
+  };
+
+  linalg::GenericOp newLinalgOp = rewriter.create<linalg::GenericOp>(
+      linalgOp.getLoc(), linalgOp.getResultTypes(), newInputs,
+      linalgOp.getOutputs(),
+      ArrayAttr::get(linalgOp->getContext(), newIndexingMaps),
+      linalgOp.getIteratorTypesAttr(), linalgOp.getDocAttr(),
+      linalgOp.getLibraryCallAttr(), bodyBuilder);
+
+  rewriter.replaceOp(linalgOp, newLinalgOp.getResults());
+
+  return success();
+}
+
+static LogicalResult tryRaiseToCollapseShape(AffineMap inputIndexingMap,
+                                             AffineMap outputIndexingMap,
+                                             Value input, Value output,
+                                             linalg::GenericOp linalgOp,
+                                             RewriterBase &rewriter) {
+  // Output shape must be smaller than input shape.
+  if (outputIndexingMap.getNumResults() >= inputIndexingMap.getNumResults())
```

Comment:
AFAICS, if you have a `linalg.generic` operation that is all parallel, then it is undefined behavior to have the output indexing map not be a permutation. Here it seems like you are looking for a case where the input is a permutation, and output is a strict projection. I think that is undefined. To extend that, I am not sure a `tensor.collapse_shape`/`tensor.expand_shape` can be represented by a `linalg.generic`. Could you post the concrete case that made you go down that path?

EDIT: I see where you are going with this based on the test below. I think a better solution would be to use a rank-reduced extract_slice here instead of trying to use a `tensor.collapse_shape`/`tensor.expand_shape`. As mentioned above, inherently you cannot use a `linalg.generic` to represent these operations, so we are going through some "undefined behavior" when we are basically dropping unit dimensions. So I'd rather use a rank-reduced extract slice.